README.md
# passport-adauth

[Passport](http://passportjs.org/) authentication strategy against Active
Directory. This module is a Passport strategy wrapper for [adauth](
https://github.com/vsimonian/node-adauth).

## Install

    $ npm install passport-adauth

## Status

[![Build Status](https://travis-ci.org/vsimonian/passport-adauth.png)](
https://travis-ci.org/vsimonian/passport-adauth)
[![Dependency Status](https://gemnasium.com/vsimonian/passport-adauth.png)](
https://gemnasium.com/vsimonian/passport-adauth)

## License

MIT. See "LICENSE" file.

## Usage

```javascript
var ADStrategy = require('passport-adauth');

passport.use(new ADStrategy({
  server: {
    url: 'ldaps://corp.example.com',
    domainDn: 'dc=example,dc=com',
    tlsOptions: {
      ca: './domain-ca.cer'
    }
  }
}));
```

* `server`: LDAP settings. These are passed directly to [adauth](
  https://github.com/vsimonian/node-adauth). See its documentation for all
  available options.
    * `url`: The LDAP URL for the domain controller e.g.
      `ldaps://corp.example.com:636`
    * `bindDn`: A user with just enough permissions to read other users'
      attributes e.g. `CN=LDAP User,OU=Users,OU=MyBusiness,DC=example,DC=com`
    * `bindCredentials`: Password for bindDn
    * `searchBase`: The base suffix for all users to narrow results e.g.
      `OU=Users,OU=MyBusiness,DC=example,DC=com`
    * `searchAttributes`: Optional array of attributes to limit what attributes
      are fetched from AD, e.g. `['displayName', 'mail']`. Defaults to
      `undefined`, i.e. fetch all attributes
    * `tlsOptions`: Optional object with options accepted by the Node.js [tls](
      http://nodejs.org/api/tls.html#tls_tls_connect_options_callback) module.
* `usernameField`: Field name where the username is found, defaults to
  _username_
* `passwordField`: Field name where the password is found, defaults to
  _password_
* `passReqToCallback`: When `true`, `req` is the first argument to the verify
  callback (default: `false`):

```js
passport.use(new ADStrategy(..., function(req, user, done) {
    ...
    done(null, user);
  }
));
```

Note: you can pass a function instead of an object as `options`, see the
[example below](#options-as-function)

### Authenticate requests

Use `passport.authenticate()`, specifying the `'adauth'` strategy, to
authenticate requests.

#### `authenticate()` options

In addition to [default authentication options](
http://passportjs.org/guide/authenticate/) the following options are available
for `passport.authenticate()`:

 * `badRequestMessage`  flash message for missing username/password (default:
   'Missing credentials')
 * `invalidCredentials`  flash message for `InvalidCredentialsError`,
   `NoSuchObjectError`, and `/no such user/i` LDAP errors (default: 'Invalid
   username/password')
 * `userNotFound`  flash message when AD returns no error but also no user
   (default: 'Invalid username/password')
 * `constraintViolation`  flash message when user account is locked (default:
   'Exceeded password retry limit, account locked')

## Express example

```javascript
var express      = require('express'),
    passport     = require('passport'),
    bodyParser   = require('body-parser'),
    ADStrategy   = require('passport-adauth');

var adOptions = {
  server: {
    url: 'ldap://corp.example.com',
    bindDn: 'CN=LDAP User,OU=Users,OU=MyBusiness,DC=example,DC=com',
    bindCredentials: 'mypassword',
    searchBase: 'OU=Users,OU=MyBusiness,DC=example,DC=com'
  }
};

var app = express();

passport.use(new ADStrategy(adOptions));

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(passport.initialize());

app.post('/login', passport.authenticate('adauth', { session: false }),
function(req, res) {
  res.send({ status: 'ok' });
});

app.listen(8080);
```

### Active Directory over SSL example

Simple example config for connecting over `ldaps://` to a server requiring some
internal CA certificate (often the case in corporations using Windows AD).

```javascript
var opts = {
  server: {
    url: 'ldaps://corp.example.com:636',
    bindDn: 'CN=LDAP User,OU=Users,OU=MyBusiness,DC=example,DC=com',
    bindCredentials: 'mypassword',
    searchBase: 'OU=Users,OU=MyBusiness,DC=example,DC=com'
    tlsOptions: {
      ca: './domain-ca.cer'
    }
  }
};
...
```

<a name="options-as-function"></a>
## Asynchronous configuration retrieval

Instead of providing a static configuration object, you can pass a function as
`options` that will take care of fetching the configuration. It will be called'
with the `req` object and a callback function having the standard `(err,
result)` signature. Notice that the provided function will be called on every
authenticate request.

```javascript
var getLDAPConfiguration = function(req, callback) {
  // Fetching things from database or whatever
  process.nextTick(function() {
    var opts = {
      server: {
        url: 'ldap://corp.example.com',
        bindDn: 'CN=LDAP User,OU=Users,OU=MyBusiness,DC=example,DC=com',
        bindCredentials: 'mypassword',
        searchBase: 'OU=Users,OU=MyBusiness,DC=example,DC=com'
      }
    };

    callback(null, opts);
  });
};

var ADStrategy = require('passport-adauth');

passport.use(new ADStrategy(getLDAPConfiguration,
  function(user, done) {
    ...
    return done(null, user);
  }
));
```


CHANGELOG.md
# Change log

## 0.1.2
### Changed
- Recommended passing CA certificate URL directly to node-adauth instead of
  using `fs.readFileSync(...)`

## 0.1.1
### Fixed
- Test in which a typo prevented it from passing

### Removed
- node 0.8 from Travis test profiles, which used an old version of npm that
  would not properly resolve the caret `^` in dependency versions
  
### Changed
- Depends on `adauth` 0.1.2 now, which returns `objectGUID` attributes in a
  string format instead of as a `Buffer` object

## 0.1.0
First working version after being forked from passport-ldapauth

### Changed
- Switched to using `adauth` instead of `ldapauth`
- Unit tests modified to create an Active Directory-like environment

test/appserver.js
var express      = require('express'),
    passport     = require('passport'),
    ADStrategy   = require('passport-adauth').Strategy,
    bodyParser   = require('body-parser');

var server = null;

var init_passport = function (opts, testopts) {
  if (testopts.no_callback === true) {
    passport.use(new ADStrategy(opts));
  } else {
    passport.use(new ADStrategy(opts, function(user, cb) {
      return cb(null, user);
    }));
  }
};

exports.start = function (opts, testopts, cb) {

  var app = express();

  init_passport(opts, testopts);

  app.use(bodyParser.json());
  app.use(passport.initialize());

  app.post('/login', passport.authenticate('adauth',
    {session: false}), function (req, res) {
    res.send({status: 'ok'});
  });

  if (typeof cb === 'function') return cb(app);
  return;
};

exports.close = function (cb) {
  if (server) server.close();
  server = null;
  if (typeof cb === 'function') return cb();
  return;
};


test/ldapserver.js
var ldap = require('ldapjs');

var authorize = function (req, res, next) {
  return next();
};

var server = null;

var db = {
  valid: {
    dn: 'CN=Valid User,OU=Users,OU=MyBusiness,DC=example,DC=com',
    attributes:  {
      userPrincipalName: 'valid@example.com',
      samAccountName: 'valid',
      primaryGroupID: 83958793,
      objectSid: [0x01,0x05,0x00,0x00,0x00,0x00,0x00,0x05,0x15,0x00,0x00,0x00,
                  0xA0,0x65,0xCF,0x7E,0x78,0x4B,0x9B,0x5F,0xE7,0x7C,0x87,0x70,
                  0x09,0x1C,0x01,0x00],
      memberOf: [
        'CN=Group 2,OU=Groups,OU=MyBusiness,DC=example,DC=com'
      ],
      objectClass: [ 'user' ],
      objectCategory: [ 'user' ]
    }
  },
  group1: {
    dn: 'CN=Group 1,OU=Groups,OU=MyBusiness,DC=example,DC=com',
    attributes: {
      name: 'Group 1',
      objectSid: [0x01,0x05,0x00,0x00,0x00,0x00,0x00,0x05,0x15,0x00,0x00,0x00,
                  0xA0,0x65,0xCF,0x7E,0x78,0x4B,0x9B,0x5F,0xE7,0x7C,0x87,0x70,
                  0x09,0x1C,0x01,0x05],
      memberOf: [],
      member: [],
      objectClass: [ 'group' ],
      objectCategory: [ 'group' ]
    }
  },
  group2: {
    dn: 'CN=Group 2,OU=Groups,OU=MyBusiness,DC=example,DC=com',
    attributes: {
      name: 'Group 2',
      objectSid: [0x01,0x05,0x00,0x00,0x00,0x00,0x00,0x05,0x15,0x00,0x00,0x00,
                  0xA0,0x65,0xCF,0x7E,0x78,0x4B,0x9B,0x5F,0xE7,0x7C,0x87,0x70,
                  0x09,0x1C,0x01,0x10],
      memberOf: [
        'CN=Group 3,OU=Groups,OU=MyBusiness,DC=example,DC=com'
      ],
      objectClass: [ 'group' ],
      objectCategory: [ 'group' ],
      member: [
        'CN=Valid User,OU=Users,OU=MyBusiness,DC=example,DC=com'
      ]
    }
  },
  group3: {
    dn: 'CN=Group 3,OU=Groups,OU=MyBusiness,DC=example,DC=com',
    attributes: {
      name: 'Group 3',
      objectSid: [0x01,0x05,0x00,0x00,0x00,0x00,0x00,0x05,0x15,0x00,0x00,0x00,
                  0xA0,0x65,0xCF,0x7E,0x78,0x4B,0x9B,0x5F,0xE7,0x7C,0x87,0x70,
                  0x09,0x1C,0x01,0x0F],
      objectClass: [ 'group' ],
      objectCategory: [ 'group' ],
      memberOf: [],
      member: [
        'CN=Group 2,OU=Groups,OU=MyBusiness,DC=example,DC=com'
      ]
    }
  }
};

exports.start = function (port, cb) {
  if (server) {
    if (typeof (cb) === 'function') return cb();
    return;
  }

  server = ldap.createServer({ strictDN: false });

  server.bind('', function(req, res, next) {
    var dn = req.dn.toString();
    if (dn !== db.valid.attributes.userPrincipalName ||
        req.credentials !== 'validpassword') {
      return next(new ldap.InvalidCredentialsError());
    }
    res.end();
  });

  server.search('', function(req, res, next) {
    var filter = req.filter.filters ?
      req.filter.filters[req.filter.filters.length - 1] : req.filter;
    if (filter.matches(db.valid.attributes)) {
      res.send(db.valid);
    } else if (filter.attribute.toLowerCase() === 'objectsid' &&
      filter.value === 'S-1-5-21-2127521184-1604012920-1887927527-83958793') {
      res.send(db.group1);
    } else if (filter.attribute.toLowerCase() === 'member' &&
      ldap.parseDN(filter.value).equals(ldap.parseDN(db.valid.dn))) {
      res.send(db.group2);
    } else if (filter.attribute.toLowerCase() === 'member' &&
      ldap.parseDN(filter.value).equals(ldap.parseDN(db.group2.dn))) {
      res.send(db.group3);
    }
    res.end();
  });

  server.listen(port, function () {
    if (typeof (cb) === 'function') return cb();
  });
};

exports.close = function (cb) {
  if (server) server.close();
  server = null;
  if (typeof (cb) === 'function') return cb();
  return;
};

if (!module.parent) {
  exports.start(1389);
}


test/strategy-test.js
var should       = require('chai').Should(),
    ADStrategy   = require('passport-adauth'),
    request      = require('supertest'),
    ldapserver   = require('./ldapserver'),
    appserver    = require('./appserver');

var LDAP_PORT = 1389;

var expressapp = null;

// Base options that are cloned where needed to edit
var BASE_OPTS = {
  server: {
    url: 'ldap://localhost:' +  LDAP_PORT.toString(),
    domainDn: 'DC=example,DC=com',
    bindDn: 'CN=LDAP User,OU=Users,OU=MyBusiness,DC=example,DC=com',
    bindCredentials: 'mypassword',
    searchBase: 'OU=Users,OU=MyBusiness,DC=example,DC=com'
  }
},
BASE_TEST_OPTS = {
  no_callback: false
};

var start_servers = function (opts, test_opts) {
  return function (cb) {
    ldapserver.start(LDAP_PORT, function () {
      appserver.start(opts, test_opts, function (app) {
        expressapp = app;
        cb();
      });
    });
  }
}

var stop_servers = function (cb) {
  appserver.close(function () {
    ldapserver.close(function () {
      cb();
    });
  });
};

describe("AD authentication strategy", function () {

  describe("by itself", function () {

    it("should export Strategy constructor directly", function (cb) {
      require('passport-adauth').should.be.a('function');
      cb();
    });

    it("should export Strategy constructor separately as well", function (cb) {
      var strategy = require('passport-adauth').Strategy;
      strategy.should.be.a('function');
      (function () {
        new strategy(BASE_OPTS);
      }).should.not.throw(Error);
      cb();
    });

    it("should be named adauth", function (cb) {
      var s = new ADStrategy(BASE_OPTS);
      s.name.should.equal('adauth');
      cb();
    });

    it("should throw an error if no arguments are provided", function (cb) {
      (function () {
        new ADStrategy();
      }).should.throw(Error);
      cb();
    });

    it("should throw an error if options are not accepted by adauth",
    function (cb) {
      var s = new ADStrategy({}, function () {});
      (function () {
        s.authenticate({body: {username: 'valid', password: 'valid'}});
      }).should.throw(Error);
      cb();
    });

    it("should initialize without a verify callback", function (cb) {
      (function () {
        new ADStrategy({server: {}})
      }).should.not.throw(Error);
      cb();
    });

  });

  describe("with basic settings", function () {

    before(start_servers(BASE_OPTS, BASE_TEST_OPTS));

    after(stop_servers);

    it("should return unauthorized if credentials are not given",
    function (cb) {
      request(expressapp)
        .post('/login')
        .send({})
        .expect(400)
        .end(cb);
    });

    it("should allow access with valid credentials", function (cb) {
      request(expressapp)
        .post('/login')
        .send({username: 'valid@example.com', password: 'validpassword'})
        .expect(200)
        .end(cb);
    });

    it("should allow access with valid credentials in query string",
    function (cb) {
      request(expressapp)
        .post('/login?username=valid@example.com&password=validpassword')
        .expect(200)
        .end(cb);
    });

    it("should return unauthorized with invalid credentials", function (cb) {
      request(expressapp)
        .post('/login')
        .send({username: 'valid@example.com', password: 'invvalid'})
        .expect(401)
        .end(cb);
    });

    it("should return unauthorized with non-existing user", function (cb) {
      request(expressapp)
        .post('/login')
        .send({username: 'nonexisting', password: 'invvalid'})
        .expect(401)
        .end(cb);
    });

    it("should return groups for user", function (cb) {
      start_servers(BASE_OPTS, BASE_TEST_OPTS)(function () {
        var req = { body: {
          username: 'valid@example.com',
          password: 'validpassword'
        } },
            s   = new ADStrategy(BASE_OPTS, function (user, done) {
              req.should.have.keys('body');
              req.body.should.have.keys(['username', 'password']);
              done(null, user);
            });

        s.success = function (user) {
          should.exist(user);
          user.userPrincipalName.should.equal('valid@example.com');
          user._groups.length.should.equal(3);
          user._groups[0].name.should.equal('Group 1');
          user._groups[1].name.should.equal('Group 2');
          user._groups[2].name.should.equal('Group 3');
          cb();
        };

        s.authenticate(req);
      });
    });
  });

  describe("without a verify callback", function () {
    before(start_servers(BASE_OPTS, {no_callback: true}));

    after(stop_servers);

    it("should still authenticate", function (cb) {
      request(expressapp)
        .post('/login')
        .send({username: 'valid@example.com', password: 'validpassword'})
        .expect(200)
        .end(cb);
    });

    it("should reject invalid event", function (cb) {
      request(expressapp)
        .post('/login')
        .send({username: 'valid@example.com', password: 'invalid'})
        .expect(401)
        .end(cb);
    });
  });

  describe("with optional options", function () {

    afterEach(stop_servers);

    it("should read given fields instead of defaults", function (cb) {
      var OPTS = JSON.parse(JSON.stringify(BASE_OPTS));
      OPTS.usernameField = 'ldapuname';
      OPTS.passwordField = 'ldappwd';

      start_servers(OPTS, BASE_TEST_OPTS)(function () {
        request(expressapp)
          .post('/login')
          .send({ldapuname: 'valid@example.com', ldappwd: 'validpassword'})
          .expect(200)
          .end(cb);
      });
    });

    it("should pass request to verify callback if defined so", function (cb) {
      var OPTS = JSON.parse(JSON.stringify(BASE_OPTS));
      OPTS.passReqToCallback = true;

      start_servers(OPTS, BASE_TEST_OPTS)(function () {
        var req = { body: {
          username: 'valid@example.com',
          password: 'validpassword',
          testkey: 1
        } },
            s   = new ADStrategy(OPTS, function (req, user, done) {
              req.should.have.keys('body');
              req.body.should.have.keys(['username', 'password', 'testkey']);
              done(null, user);
            });

        s.success = function (user) {
          should.exist(user);
          user.userPrincipalName.should.equal('valid@example.com');
          cb();
        };

        s.authenticate(req);
      });
    });
  });

  describe("with options as function", function () {
    var OPTS = JSON.parse(JSON.stringify(BASE_OPTS));
    OPTS.usernameField = 'cb_uname';
    OPTS.passwordField = 'cb_pwd';

    var opts = function (cb) {
      process.nextTick(function () {
        cb(null, OPTS);
      });
    };

    before(start_servers(opts, BASE_TEST_OPTS));
    after(stop_servers);

    it("should use the options returned from the function", function (cb) {
      request(expressapp)
        .post('/login')
        .send({cb_uname: 'valid@example.com', cb_pwd: 'validpassword'})
        .expect(200)
        .end(cb);
    });

    it("should not allow login if using wrong fields", function (cb) {
      request(expressapp)
        .post('/login')
        .send({username: 'valid@example.com', password: 'validpassword'})
        .expect(400)
        .end(cb);
    });
  });

  describe("with options as function returning dynamic sets", function () {
    var OPTS = JSON.parse(JSON.stringify(BASE_OPTS));
    OPTS.usernameField = 'first_uname';
    OPTS.passwordField = 'first_pwd';

    var OPTS2 = JSON.parse(JSON.stringify(BASE_OPTS));
    OPTS2.usernameField = 'second_uname';
    OPTS2.passwordField = 'second_pwd';

    var opts = function (req, cb) {
      process.nextTick(function () {
        if (req.body.set == 'first') {
          cb(null, OPTS);
        } else {
          cb(null, OPTS2);
        }
      });
    };

    before(start_servers(opts, BASE_TEST_OPTS));
    after(stop_servers);

    it("should use the first set options returned from the function",
    function (cb) {
      request(expressapp)
        .post('/login')
        .send({
          first_uname: 'valid@example.com',
          first_pwd: 'validpassword',
          set: 'first'
        })
        .expect(200)
        .end(cb);
    });

    it("should not allow first set login if using wrong fields", function (cb) {
      request(expressapp)
        .post('/login')
        .send({
          second_uname: 'valid@example.com',
          second_pwd: 'validpassword',
          set: 'first'
        })
        .expect(400)
        .end(cb);
    });

    it("should use the second set options returned from the function",
    function (cb) {
      request(expressapp)
        .post('/login')
        .send({
          second_uname: 'valid@example.com',
          second_pwd: 'validpassword',
          set: 'second'
        })
        .expect(200)
        .end(cb);
    });

    it("should not allow second set login if using wrong fields",
    function (cb) {
      request(expressapp)
        .post('/login')
        .send({
          first_uname: 'valid@example.com',
          first_pwd: 'validpassword',
          set: 'second'
        })
        .expect(400)
        .end(cb);
    });
  });
});


lib/passport-adauth/strategy.js
"use strict";

/**
 * Passport wrapper for adauth
 */
var passport = require('passport-strategy'),
    ADAuth   = require('adauth'),
    util     = require('util');

/**
 * Strategy constructor
 *
 * The AD authentication strategy authenticates requests based on the
 * credentials submitted through an HTML-based login form.
 *
 * Applications may supply a `verify` callback which accepts `user` object
 * and then calls the `done` callback supplying a `user`, which should be set
 * to `false` if user is not allowed to authenticate. If an exception occured,
 * `err` should be set.
 *
 * Options:
 *
 * - `server`             options for adauth, see
 *                        https://github.com/vsimonian/node-adauth
 *
 * - `usernameField`      field name where the username is found, defaults to
 *                        _username_
 *
 * - `passwordField`      field name where the password is found, defaults to
 *                        _password_
 *
 * - `passReqToCallback`  when `true`, `req` is the first argument to the verify
 *                        callback (default: `false`)
 *
 * Options can be also given as function that accepts a callback end calls it
 * with error and options arguments. Notice that the callback is executed on
 * every authenticate call.
 *
 * Example:
 *
 *     var ADStrategy = require('passport-adauth').Strategy;
 *     passport.use(new ADStrategy({
 *         server: {
 *           url: 'ldaps://corp.example.com:636',
 *           bindDn: 'CN=LDAP User,OU=Users,OU=MyBusiness,DC=example,DC=com',
 *           bindCredentials: 'mypassword',
 *           searchBase: 'OU=Users,OU=MyBusiness,DC=example,DC=com',
 *           tlsOptions: {
 *             { ca: './domain-ca.cer' }
 *           }
 *         }
 *       },
 *       function(user, done) {
 *         return cb(null, user);
 *       }
 *     ));
 */
var Strategy = function(options, verify) {
  // We now accept function as options as well so we cannot assume anymore
  // that a call with a function parameter only would have skipped options
  // and just provided a verify callback
  if (!options) {
    throw new Error('AD authentication strategy requires options');
  }

  this.options    = null;
  this.getOptions = null;

  if (typeof options === 'object') {
    this.options = setDefaults(options);
  } else if (typeof options === 'function') {
    this.getOptions = options;
  }

  passport.Strategy.call(this);

  this.name   = 'adauth';
  this.verify = verify;
};

util.inherits(Strategy, passport.Strategy);

/**
 * Add default values to options
 *
 * @param options
 * @returns {*}
 */
var setDefaults = function(options) {
  options.usernameField || (options.usernameField = 'username');
  options.passwordField || (options.passwordField = 'password');
  return options;
};

/**
 * Get value for given field from given object. Taken from passport-local,
 * copyright 2011-2013 Jared Hanson
 */
var lookup = function (obj, field) {
  var i, len, chain, prop;
  if (!obj) { return null; }
  chain = field.split(']').join('').split('[');
  for (i = 0, len = chain.length; i < len; i++) {
    prop = obj[chain[i]];
    if (typeof(prop) === 'undefined') { return null; }
    if (typeof(prop) !== 'object') { return prop; }
    obj = prop;
  }
  return null;
};

/**
 * Verify the outcome of caller verify function - even if authentication (and
 * usually authorization) is taken care by LDAP there may be reasons why
 * a verify callback is provided, and again reasons why it may reject login
 * for a valid user.
 */
var verify = function() {
  // Callback given to user given verify function.
  return function(err, user, info) {
    if (err)   return this.error(err);
    if (!user) return this.fail(info);
    return this.success(user, info);
  }.bind(this);
};

/**
 * Options (passed when calling `passport.authenticate()`):
 * - `badRequestMessage`    flash message for missing username/password
 *                          (default: 'Missing credentials')
 *
 * - `invalidCredentials`   flash message for InvalidCredentialsError,
 *                          NoSuchObjectError, and /no such user/ LDAP errors
 *                          (default: 'Invalid username/password')
 *
 * - `userNotFound`         flash message when AD returns no error but also no
 *                          user
 *                          (default: 'Invalid username/password')
 *
 * - `constraintViolation`  flash message when user account is locked
 *                          (default: 'Exceeded password retry limit, account
 *                          locked')
 */
var handleAuthentication = function(req, options) {
  var username, password, ad;
  options || (options = {});

  username = lookup(req.body, this.options.usernameField) || lookup(req.query, this.options.usernameField);
  password = lookup(req.body, this.options.passwordField) || lookup(req.query, this.options.passwordField);

  if (!username || !password) {
    return this.fail({message: options.badRequestMessage || 'Missing credentials'}, 400);
  }

  ad = new ADAuth(this.options.server);
  ad.authenticate(username, password, function(err, user) {
    ad.close(function(){}); // We don't care about the closing
    if (err) {
      // Invalid credentials / user not found are not errors but login failures
      if (err.name === 'InvalidCredentialsError' || err.name === 'NoSuchObjectError' || (typeof err === 'string' && err.match(/no such user/i))) {
        return this.fail({message: options.invalidCredentials || 'Invalid username/password'}, 401);
      }
      if (err.name === 'ConstraintViolationError'){
        return this.fail({message: options.constraintViolation || 'Exceeded password retry limit, account locked'}, 401);
      }
      // Other errors are (most likely) real errors
      return this.error(err);
    }

    if (!user) return this.fail({message: options.userNotFound || 'Invalid username/password'}, 401);

    // Execute given verify function
    if (this.verify) {
      if (this.options.passReqToCallback) {
        return this.verify(req, user, verify.call(this));
      } else {
        return this.verify(user, verify.call(this));
      }
    } else {
      return this.success(user);
    }
  }.bind(this));
};

/**
 * Authenticate the request coming from a form or such.
 */
Strategy.prototype.authenticate = function(req, options) {
  if ((typeof this.options === 'object') && (!this.getOptions)) {
    return handleAuthentication.call(this, req, options);
  }
  var callback = function(err, configuration) {
    if (err) return this.fail(err);

    this.options = setDefaults(configuration);
    handleAuthentication.call(this, req, options);
  };
  // Added functionality: getOptions can accept now up to 2 parameters
  if (this.getOptions.length ===1) { // Accepts 1 parameter, backwards compatibility
    this.getOptions(callback.bind(this));
  } else { // Accepts 2 parameters, pass request as well
    this.getOptions(req, callback.bind(this));
  }
};

module.exports = Strategy;
